src/storage/s3.rs

```diff
@@ -1,50 +1,77 @@
 use failure::Error;
 use futures_util::{
     future::TryFutureExt,
     stream::{FuturesUnordered, StreamExt},
 };
 use log::warn;
-use once_cell::sync::Lazy;
-use rusoto_core::region::Region;
+use rusoto_core::{region::Region, RusotoError};
 use rusoto_credential::DefaultCredentialsProvider;
-use rusoto_s3::{GetObjectRequest, PutObjectRequest, S3Client, S3};
+use rusoto_s3::{
+    DeleteObjectsRequest, GetObjectError, GetObjectRequest, ListObjectsV2Request, ObjectIdentifier,
+    PutObjectRequest, S3Client, S3,
+};
 use std::{convert::TryInto, io::Write};
 use tokio::runtime::Runtime;
 
-#[cfg(test)]
-mod test;
-#[cfg(test)]
-pub(crate) use test::TestS3;
-
-pub(crate) static S3_BUCKET_NAME: &str = "rust-docs-rs";
-pub(crate) static S3_RUNTIME: Lazy<Runtime> =
-    Lazy::new(|| Runtime::new().expect("Failed to create S3 runtime"));
-
-pub(crate) struct S3Backend {
-    pub client: S3Client,
+pub(super) struct S3Backend {
+    client: S3Client,
+    runtime: Runtime,
     bucket: String,
+    #[cfg(test)]
+    temporary: bool,
 }
 
 impl S3Backend {
-    pub(crate) fn new(client: S3Client, bucket: &str) -> Self {
-        Self {
-            client,
-            bucket: bucket.into(),
+    pub(super) fn new(client: S3Client, config: &Config) -> Result<Self, Error> {
+        let runtime = Runtime::new()?;
+
+        // Create the temporary S3 bucket during tests.
+        if config.s3_bucket_is_temporary {
+            if cfg!(not(test)) {
+                panic!("safeguard to prevent creating temporary buckets outside of tests");
+            }
+
+            runtime
+                .handle()
+                .block_on(client.create_bucket(rusoto_s3::CreateBucketRequest {
+                    bucket: config.s3_bucket.clone(),
+                    ..Default::default()
+                }))?;
         }
+
+        Ok(Self {
+            client,
+            runtime,
+            bucket: config.s3_bucket.clone(),
+            #[cfg(test)]
+            temporary: config.s3_bucket_is_temporary,
+        })
     }
 
     pub(super) fn get(&self, path: &str, max_size: usize) -> Result<Blob, Error> {
-        S3_RUNTIME.handle().block_on(async {
-            let res = self
+        self.runtime.handle().block_on(async {
+            let response = self
                 .client
                 .get_object(GetObjectRequest {
                     bucket: self.bucket.to_string(),
                     key: path.into(),
                     ..Default::default()
                 })
-                .await?;
+                .await;
+
+            let res = match response {
+                Ok(res) => res,
+                Err(RusotoError::Service(GetObjectError::NoSuchKey(_))) => {
+                    return Err(super::PathNotFoundError.into());
+                }
+                Err(RusotoError::Unknown(http)) if http.status == 404 => {
+                    return Err(super::PathNotFoundError.into());
+                }
```

Comment:
Maybe these errors should be logged with more details than just the `PathNotFoundError`
```suggestion
                Err(RusotoError::Service(GetObjectError::NoSuchKey(key))) => {
                    log::error!("S3 didn't recognize a key: {:?}", key);
                    return Err(super::PathNotFoundError.into());
                }
                Err(RusotoError::Unknown(http)) if http.status == 404 => {
                    log::error!("Attempted to get an unknown file from S3: {:?}", http);
                    return Err(super::PathNotFoundError.into());
                }
```

Reply:
I mean, they're just 404. Not sure why the extra logging would be useful.